### ISSUE TITLE
Exception if duration is non-multiple of timestep

### ISSUE THREAD

**Issue by astoeckel** [SpiNNakerManchester/sPyNNaker#149]
In contrast to all other platform sPyNNaker crashes if the simulation duration is not a multiple of the simulation timestep:

         "The runtime and machine time step combination result in "
         spinn_front_end_common.utilities.exceptions.ConfigurationException:
         The runtime and machine time step combination result in a factional number
         of machine runable time steps and therefore spinnaker cannot determine
         how many to run for

sPyNNaker should round the duration up to the next timestep instead of throwing an exception for compatibility with other platforms.

See also: https://github.com/hbp-sanncs/pynnless/commit/d5b6fdab82099b96d984c00c63beecca2a63d5f8

**Comment by AlexRast** [comment]
In this instance I would argue that other plaforms are making a fairly dangerous 
assumption about what is intended if the user sets a duration which isn't a 
multiple of the timestep. There is no easy way, I think, to divine what the 
user's intent (or for that matter lack of intent, it could simply be 
inadvertency on the user's part) would be in asking for essentially incompatible 
durations and timesteps. My feeling is that the duration should always be in 
multiples of the timestep.

If other systems are silently executing without error by making some sort of 
default assumption in cases of specification ambiguity, this will make 
simulations prone to subtle, difficult to debug errors or inaccuracies in 
future. This particular situation is, I think, a good example of the pitfalls of 
blindly copying parameters without considering what they mean or how they may 
interact.

I recommend that we should indeed suggest to other HBP platform developers that 
setting a duration not a multiple of the simulation timestep should raise an 
error. This should be true even for asynchronous simulations with no physical 
timestep: real systems will always have finite time resolution, which is what 
the timestep parameter should represent. Perhaps the term "timestep" is poor and 
ought to be changed to "time resolution" but this is a separate discussion.

**Comment by apdavison** [comment]
NEST emits a warning in this situation. I think this would be the best solution for sPyNNaker as well, as it ensures the user is aware of potential problems while still allowing maximum flexibility for the user.

**Comment by astoeckel** [comment]
I second Andrews proposal, just let me add a few more thoughts:

* If the function would only accept multiples of the time step, such a restriction should be made clear on an API level via the data type. For multiples of a time step, the correct data type would be an integer (expressing the number of time steps), not a float. Due to limited precision it is impossible to check correctly whether a floating point value is a multiple of something, e.g.

```python
a = 1.0 / 3.0; c = 100000; (c * a) / a == c # outputs "False"
```

* There will always be differences in the capabilities of the different PyNN backends. If the user runs into such a limitation, the platform should inform about this by issuing a warning (exactly once per limitation and execution) and work around this as best as possible. It is almost never an option to throw an exception. This will force users to add platform-specific workaround to their own code (just as I collected them in PyNNLess, as of now summing up to 1200 lines of if-then-else) and works against what PyNN is actually about -- providing a coherent interface that allows you to run the same script on multiple, conceptually different platforms. sPyNNaker is by far the most "nitpicky" backend, although it is conceptually more flexible than the physical model.

**Comment by rowleya** [comment]
If you set the timestep arbitrarily, this is true.  This is even more complicated in our case anyway, as both the timestep and the run time will be converted into integers of microseconds.  I am happy to add some warnings to this, but you are right that taking a guess and warning about what we have guessed is a good general solution.  Note though that our guess might not be the same guess as that taken by other systems; for example, until recently, we recorded voltage after the timestep update rather than before.  It was only when we compared our output to NEST that we found that this was not "standard" (whatever that means).  

There will always be differences like this until the system has been used for a long time and compared to other systems (the software as it currently exists is only about 1 year old now).  We are always happy for you to report these differences though, and we will update them when we get around to it (we are but a small team of developers).

**Comment by AlexRast** [comment]
On 22/10/15 20:58, Andreas Stöckel wrote:
> I second Andrews proposal, just let me add a few more thoughts:
>
>   * If the function would only accept multiples of the time step, such a
>     restriction should be made clear on an API level via the data type. For
>     multiples of a time step, the correct data type would be an integer
>     (expressing the number of time steps), not a float. Due to limited precision
>     it is impossible to check correctly whether a floating point value is a
>     multiple of something, e.g.
>
> a=  1.0  /  3.0; c=  100000; (c*  a)/  a==  c# outputs "False"

The SpiNNaker backend code is strongly typed and based on a specific set of 
types that can be determined by looking through the headers. It is important, 
however, to be mindful of the fact that PyNN itself is written in Python, which 
isn't strongly typed.
>
>   * There will always be differences in the capabilities of the different PyNN
>     backends. If the user runs into such a limitation, the platform should
>     inform about this by issuing a warning (exactly once per limitation and
>     execution) and work around this as best as possible. It is almost never an
>     option to throw an exception. This will force users to add platform-specific
>     workaround to their own code (just as I collected them in PyNNLess, as of
>     now summing up to 1200 lines of if-then-else) and works against what PyNN is
>     actually about -- providing a coherent interface that allows you to run the
>     same script on multiple, conceptually different platforms. sPyNNaker is by
>     far the most "nitpicky" backend, although it is conceptually more flexible
>     than the physical model.

Some things the platform may be able to warn about, although there are cases 
where something flatly can't be supported (e.g. if the user asks for a neural 
model that isn't available on the current platform), and I think in these cases 
it is appropriate to issue an error. Is there a way to make sure warning 
messages don't pass by in a sea of informational messages?

One upshot is that because of issues like this, where there is an ambiguity in 
the definition, I think there may need to be a "reference" platform defined 
in/by PyNN which is to be taken as the standard execution model. As Andrew 
Rowley has pointed out if 2 different platforms make opposite or conflicting 
assumptions about what to do, in cases where some sort of workaround is 
possible, comparing results will be very difficult indeed.

In a larger sense, as ever-more users start using PyNN generally and SpiNNaker 
in particular, there are going to be more unusual, unanticipated use cases which 
expose underlying limitations in the design. If we gather together lists of 
these issues for support in a future release, while documenting what the 
limitations are in the current release that seems like an acceptable approach.

**Comment by astoeckel** [comment]
> It is important, however, to be mindful of the fact that PyNN itself is written in Python, which isn't strongly typed.

I'm aware of that ;-) And the inability to specify a statically-typed interface in Python is actually a significant issue for such a large-scale project as PyNN and its backend in my opinion.

> Some things the platform may be able to warn about, although there are cases where something flatly can't be supported (e.g. if the user asks for a neural model that isn't available on the current platform), and I think in these cases it is appropriate to issue an error.

Yes, I also think that in such cases exceptions are entirely appropriate. I was referring to those small issues for which a simple workaround exists.

> Is there a way to make sure warning messages don't pass by in a sea of informational messages?

In PyNNLess I hacked this together by storing all warnings in a set and displaying them at the latest possible moment. This makes sure that each warning class is shown exactly once per execution, however, one must refrain from parametrising the messages (or index the set entries by a message type instead of their string). I don't know whether the Python logging system already provides something like this. A side note here: I actually have to throw away most of the log output from the Heidelberg system in PyNNLess (via these unearthly _redirect/_unredirect_io methods), because stderr gets so large, that the neuromorphic compute platform hangs when copying the results back. So displaying few concise warnings is really important when going down this path.

> I think there may need to be a "reference" platform defined in/by PyNN which is to be taken as the standard execution model.

That might be a solution, a set of backend-independent unit-tests for edge cases could also helpful.

> If we gather together lists of these issues for support in a future release, while documenting what the limitations are in the current release that seems like an acceptable approach.

Feel free to have a look at PyNNLess (https://github.com/hbp-sanncs/pynnless/blob/master/pynnless/pynnless.py) where I collected some differences. The differences may not look too severe, but most of them took me hours to hunt down. However, I think I reported all of them to the corresponding backend developers.
